### DISCUSSION TITLE
Marquez `v2.UI`

### DISCUSSION THREAD

**Discussion by wslulciuc** [MarquezProject/marquez#2823] [Ideas]
Let's chat about everything Marquez `v2.UI`!

**Comment by phixMe** [comment]
- Search first class citizen
- Moving away from `jobs` being the main page.
- Smaller drawer
- Adding charts
<img width="1455" alt="image" src="https://github.com/MarquezProject/marquez/assets/7514204/436649e8-9382-45dc-86c1-5405dfcba98a">

**Comment by wslulciuc** [comment]
I would take the `dashboard` view (which is already a huge improvement!) a step further and add:
1. A `view` for OpenLineage events collected within the past `8h`, `24h`, `7days` broken down by type: `COMPLETED`, `FAILED`, etc (with percentage breakdown)
2. A `view` for # of data sources
3. A `view` for # of datasets
4. A `view` for # of datasets with schemas (% without schemas)
5. A `view` for # of jobs
6. A `view` for # of runs (with percentage breakdown by run state)
7. A **timeline** `view` of failed runs within the past `8h`, `24h`, `7days` detailing: job info, job facets, runID, run facets, etc
8. A **timeline** `view` of metadata changes (`CMC`=change metadata capture) within the past `8h`, `24h`, `7days`: "_run `R` completed in `X`.minutes and created dataset version `V`_", "_new column has been added to dataset `D` by run `R` and new version `V` has be created_", etc

**Reply by davidsharp7** [reply]
Ability to update dataset/dataset field descriptions.  Being able to add "notes" to datasets/ jobs

**Reply by wslulciuc** [reply]
Yes! Actually, we were thinking about defaulting to markdown for notes! And, for descriptions, we should absolutely allow users to update via the UI, but we also need to handle the case where descriptions are provided via OpenLineage.

**Reply by davidsharp7** [reply]
I was thinking we could extend the model to have the "technical" description from open lineage then have a "business" description that users could update.  Best of both worlds then.  I will have more of a think about it though.

**Reply by wslulciuc** [reply]
I think _notes_ for datasets/jobs can cover the _technical_ details as well therefore avoiding description inconsistencies that might emerge between user entered vs OpenLineage provided descriptions (ex: would `users.name` need two descriptions? but I can see `users` table having a _note_ for its usage, etc). I feel users should annotate when descriptions are missing, but the source of truth still needs to be the what's provided in the OpenLineage event. I'm sure there's somewhere in between we can agree on.

**Reply by batwalrus76** [reply]
I think making the tables for jobs, etc. able to filter by name or time range and reorder by a column (ascending/descending order).  This would be helpful when you have a lot of ioba, events, or datasets and you don’t want to page through to find something you sort of know what you are looking for. Also including the page number and possibly skip to start or end?

**Reply by wslulciuc** [reply]
@batwalrus76 I couldn't agree more! Mind creating an issue for your table filtering suggestion using the label [`v2.UI`](https://github.com/MarquezProject/marquez/labels/v2.UI) label?

**Reply by wslulciuc** [reply]
@phixMe Below, I expand a bit more and breakdown my thoughts on  the `dashboard` views mentioned above:

1. **Dashboard:** _OpenLineage events within the past `8h`, `24h`, `7days` broken down run state_

* As a user, I want the option to select an OL event type, then be provided with a view to further explore the details of jobs with that given run state. The jobs metadata would be listed in a view (just below the graph?) with the following details:
  * `job.namespace`
  * `job.name`
  * `job.type` (_batch_, _streaming_, etc)
  * `job.updatedAt`
  * `job.platform` (_airflow_, _spark_, etc)
  * **#** of `job.inputs`
  * **#** of `job.outputs`
  * **#** of `job.inputs`
  * _current_ `run.state` (with `run.transitionedAt`)
  * **#** of last **X** runs (or graph of run durations)

2. **Timeline**: View of metadata changes (CMC=change metadata capture) within the past 8h, 24h, 7days captured by Marquez.

* As a user, I want a timeline view of metadata changes captured by Marquez. The timeline view would be chronological with the following details `CMC` events:
  * Sources
    * `source.name` **added** at `source.createdAt`
    * `dataset.name` **added** to `source.name` at `source.updatedAt`
    * `source.name` **deleted** at `source.createdAt`
  * Datasets
    * `dataset.name` **added** at `dataset.createdAt` (by `run.id`)
    * `dataset.name` schema **modified** **added** at `dataset.updatedAt` (by `run.id`); **new version** `dataset.version` **created** caused by one of:
      * `dataset.field.name` **added**
      * `dataset.field.name` **deleted**
      * `dataset.field.type` **changed**
    * `dataset.name` **deleted** (_hard_ / _soft_ deletion)
  * Jobs
      * `job.name` **added** at `job.createdAt` 
      * `job.name` source code **modified** at `job.updatedAt`; **new version** `job.version` **created** 
      * `job.name` inputs **modified** at `job.updatedAt`; **new version** `job.version` **created** (list inputs **added** / **removed**)
      * `job.name` outputs **modified** at `job.updatedAt`; **new version** `job.version` **created** (list output **added** / **removed**)
  * Runs
    * `run.id` **`STARTED`|`COMPLETED`|`FAILED`** at `run.transitionedAt`; display `run.duration`

**Comment by wslulciuc** [comment]
We've merged https://github.com/MarquezProject/marquez/pull/2825 with the new layout for the `search bar` and a new (re)focus on driving metadata exploration via _search_! More to come 🚀

**Reply by phixMe** [reply]
![image](https://github.com/MarquezProject/marquez/assets/7514204/d8236751-b4ca-43a2-bb73-ba49d331ef07)
![image](https://github.com/MarquezProject/marquez/assets/7514204/f987d5b7-8fb7-4945-838f-892dd4a996df)
![image](https://github.com/MarquezProject/marquez/assets/7514204/1c361292-517f-4252-9ced-691d7c0d9cd8)

**Comment by batwalrus76** [comment]
This is really not just a UI question but it could greatly change the graph visualizations (table and column lineage).  Should there be a way to version these graphs? I recognized the dataset schemas have versions but what happens if you modify say a spark workflow (especially in Databricks where it is super easy) and change the jobs.  I am pretty sure the graph will show both the old jobs and the new ones.  This is especially confusing when you change input/output tables.  Then you end up with a graph that has old and new state.  I also think the change in lineage can be very helpful, especially if you are doing forensics and you need a snapshot of the graph at a certain time.  There of course could be infinity ways to display this history, but really any of them would should be good enough to convey change over time.

**Reply by wslulciuc** [reply]
@batwalrus76 In Marquez, we always return the _current_ lineage graph. That is, if any inputs/outputs change for a job, the graph will be updated to reflect the change. Marquez does keep track of all dataset and job changes as _versions_, and lineage graph _snapshot_ are coming ;), see [`Proposal: Run-level Lineage in Marquez API`](https://github.com/MarquezProject/marquez/blob/main/proposals/2117-marquez-over-time.md) and https://github.com/MarquezProject/marquez/issues/2772

**Reply by batwalrus76** [reply]
I was thinking along the lines of what the proposal stated and your write-up in the above mentioned issue makes total sense.

**Comment by phixMe** [comment]
Here as some initial rendering on where we are thinking of heading in the next month or so encompassing the change events we wish to log.
![image](https://github.com/MarquezProject/marquez/assets/7514204/49d89d7a-215f-4774-a391-eb2e59c22e37)
